azurerm/resource_arm_application_gateway.go

```diff
@@ -0,0 +1,1691 @@
+			"id":                  *config.ID,
+			"name":                *config.Name,
+			"protocol":            string(config.ApplicationGatewayProbePropertiesFormat.Protocol),
+			"path":                *config.ApplicationGatewayProbePropertiesFormat.Path,
+			"host":                *config.ApplicationGatewayProbePropertiesFormat.Host,
+			"interval":            int(*config.ApplicationGatewayProbePropertiesFormat.Interval),
+			"timeout":             int(*config.ApplicationGatewayProbePropertiesFormat.Timeout),
+			"unhealthy_threshold": int(*config.ApplicationGatewayProbePropertiesFormat.UnhealthyThreshold),
+		}
+
+		result = append(result, settings)
+	}
+
+	return result
+}
+
+func flattenApplicationGatewayRequestRoutingRules(rules *[]network.ApplicationGatewayRequestRoutingRule) []interface{} {
+	result := make([]interface{}, 0, len(*rules))
+
+	for _, config := range *rules {
+		listener := map[string]interface{}{
+			"id":                 *config.ID,
+			"name":               *config.Name,
+			"rule_type":          string(config.ApplicationGatewayRequestRoutingRulePropertiesFormat.RuleType),
+			"http_listener_id":   *config.ApplicationGatewayRequestRoutingRulePropertiesFormat.HTTPListener.ID,
+			"http_listener_name": path.Base(*config.ApplicationGatewayRequestRoutingRulePropertiesFormat.HTTPListener.ID),
+		}
+
+		if config.ApplicationGatewayRequestRoutingRulePropertiesFormat.BackendAddressPool != nil {
+			listener["backend_address_pool_name"] = path.Base(*config.ApplicationGatewayRequestRoutingRulePropertiesFormat.BackendAddressPool.ID)
+			listener["backend_address_pool_id"] = *config.ApplicationGatewayRequestRoutingRulePropertiesFormat.BackendAddressPool.ID
+		}
+
+		if config.ApplicationGatewayRequestRoutingRulePropertiesFormat.BackendHTTPSettings != nil {
+			listener["backend_http_settings_name"] = path.Base(*config.ApplicationGatewayRequestRoutingRulePropertiesFormat.BackendHTTPSettings.ID)
+			listener["backend_http_settings_id"] = *config.ApplicationGatewayRequestRoutingRulePropertiesFormat.BackendHTTPSettings.ID
+		}
+
+		if config.ApplicationGatewayRequestRoutingRulePropertiesFormat.URLPathMap != nil {
+			listener["url_path_map_name"] = path.Base(*config.ApplicationGatewayRequestRoutingRulePropertiesFormat.URLPathMap.ID)
+			listener["url_path_map_id"] = *config.ApplicationGatewayRequestRoutingRulePropertiesFormat.URLPathMap.ID
+		}
+
+		result = append(result, listener)
+	}
+
+	return result
+}
+
+func flattenApplicationGatewayURLPathMaps(pathMaps *[]network.ApplicationGatewayURLPathMap) []interface{} {
+	result := make([]interface{}, 0, len(*pathMaps))
+
+	for _, config := range *pathMaps {
+		pathMap := map[string]interface{}{
+			"id":   *config.ID,
+			"name": *config.Name,
+		}
+
+		if config.ApplicationGatewayURLPathMapPropertiesFormat.DefaultBackendAddressPool != nil {
+			pathMap["default_backend_address_pool_name"] = path.Base(*config.ApplicationGatewayURLPathMapPropertiesFormat.DefaultBackendAddressPool.ID)
```

Comment:
we're assuming here the name will be at the end - can we switch this to using the `parseAzureResourceID` function to parse the values out of the URL to guarantee that this is the case?